File: src/main/java/com/timepath/hl2/ExternalConsole.java
package com.timepath.hl2;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.script.Invocable;
import javax.script.ScriptEngine;
import javax.script.ScriptEngineManager;
import javax.script.ScriptException;
import javax.swing.*;
import javax.swing.text.DefaultCaret;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.WindowAdapter;
import java.awt.event.WindowEvent;
import java.io.*;
import java.net.InetAddress;
import java.net.Socket;
import java.util.logging.Level;
import java.util.logging.Logger;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * http://www.perkin.org.uk/posts/how-to-fix-stdio-buffering.html
 *
 * @author TimePath
 */
@SuppressWarnings("serial")
public class ExternalConsole extends JFrame {

    private static final Logger LOG = Logger.getLogger(ExternalConsole.class.getName());
    private static final Pattern regex = Pattern.compile("(\\S+)\\s*[(]\\s*(\\S*)\\s*[)].*");
    @NotNull
    private final JTextField input;
    @NotNull
    private final JTextArea output;
    private ScriptEngine engine = initScriptEngine();
    private PrintWriter pw;
    private Socket sock;

    protected ExternalConsole() {
        output = new JTextArea();
        output.setFont(new Font("Monospaced", Font.PLAIN, 15));
        output.setEnabled(false);
        @NotNull DefaultCaret caret = (DefaultCaret) output.getCaret();
        caret.setUpdatePolicy(DefaultCaret.ALWAYS_UPDATE);
        @NotNull JScrollPane jsp = new JScrollPane(output);
        jsp.setHorizontalScrollBarPolicy(ScrollPaneConstants.HORIZONTAL_SCROLLBAR_AS_NEEDED);
        jsp.setVerticalScrollBarPolicy(ScrollPaneConstants.VERTICAL_SCROLLBAR_ALWAYS);
        input = new JTextField();
        input.setEnabled(false);
        input.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent e) {
                if (pw == null) {
                    return;
                }
                getOutput().append("] ");
                pw.println(input.getText());
                input.setText("");
            }
        });
        @NotNull JMenuBar jmb = new JMenuBar();
        setJMenuBar(jmb);
        @NotNull JMenu fileMenu = new JMenu("File");
        jmb.add(fileMenu);
        @NotNull JMenuItem reload = new JMenuItem("Reload script");
        fileMenu.add(reload);
        reload.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent e) {
                engine = initScriptEngine();
            }
        });
        setTitle("External console");
        //        setAlwaysOnTop(true);
        //        setUndecorated(true);
        setPreferredSize(new Dimension(800, 600));
        addWindowListener(new WindowAdapter() {
            @Override
            public void windowClosing(WindowEvent e) {
                if (sock != null) {
                    try {
                        sock.close();
                    } catch (IOException ex) {
                        LOG.log(Level.SEVERE, null, ex);
                    }
                }
                dispose();
            }
        });
        getContentPane().add(jsp, BorderLayout.CENTER);
        getContentPane().add(input, BorderLayout.SOUTH); // TODO: work out better way of sending input
        pack();
    }

    @NotNull
    public static String exec(String cmd, @Nullable CharSequence breakline) {
        @NotNull StringBuilder sb = new StringBuilder();
        try {
            @NotNull Socket sock = new Socket(InetAddress.getByName(null), 12345);
            @NotNull PrintWriter pw = new PrintWriter(sock.getOutputStream(), true);
            @NotNull BufferedReader in = new BufferedReader(new InputStreamReader(sock.getInputStream()));
            pw.println(cmd);
            if (breakline != null) {
                in.readLine(); // first line is echoed
                String line;
                while ((line = in.readLine()) != null) {
                    sb.append(line).append('\n');
                    if (line.contains(breakline)) {
                        break;
                    }
                }
            }
            sock.close();
        } catch (IOException ex) {
            Logger.getLogger(ExternalConsole.class.getName()).log(Level.SEVERE, null, ex);
        }
        return sb.toString();
    }

    public static void main(String... args) throws Exception {
        @NotNull ExternalConsole ec = new ExternalConsole();
        ec.connect(12345);
        ec.setVisible(true);
    }

    public static void setErr(@NotNull final InputStream s) {
        new Thread(new Runnable() {
            @Override
            public void run() {
                try {
                    @NotNull BufferedReader in = new BufferedReader(new InputStreamReader(s));
                    String line;
                    while ((line = in.readLine()) != null) {
                        System.err.println(line);
                    }
                    System.err.println("Stopped reading stderr");
                } catch (IOException ex) {
                    Logger.getLogger(ExternalConsole.class.getName()).log(Level.SEVERE, null, ex);
                }
            }
        }).start();
    }

    private ScriptEngine initScriptEngine() {
        @NotNull ScriptEngineManager factory = new ScriptEngineManager();
        ScriptEngine scriptEngine = factory.getEngineByName("JavaScript");
        //        Bindings bindings = engine.createBindings();
        //        bindings.put("loadTime", new Date());
        scriptEngine.getContext().setWriter(pw);
        try {
            scriptEngine.eval(new FileReader("extern.js"));
        } catch (@NotNull ScriptException | FileNotFoundException ex) {
            Logger.getLogger(ExternalConsole.class.getName()).log(Level.SEVERE, null, ex);
        }
        return scriptEngine;
    }

    /**
     * @return the output
     */
    @NotNull
    protected JTextArea getOutput() {
        return output;
    }

    protected void connect(int port) throws IOException {
        sock = new Socket(InetAddress.getByName(null), port);
        setIn(sock.getInputStream());
        setOut(sock.getOutputStream());
    }

    void setIn(@Nullable final InputStream s) {
        output.setEnabled(s != null);
        new Thread(new Runnable() {
            @Override
            public void run() {
                try {
                    @NotNull BufferedReader in = new BufferedReader(new InputStreamReader(s));
                    String line;
                    while ((line = in.readLine()) != null) {
                        update(line);
                    }
                    System.err.println("Stopped reading stdout");
                } catch (IOException ex) {
                    Logger.getLogger(ExternalConsole.class.getName()).log(Level.SEVERE, null, ex);
                }
            }
        }).start();
    }

    void update(@NotNull String str) {
        parse(str);
        appendOutput(str);
    }

    private void appendOutput(String str) {
        output.append(str + '\n');
    }

    protected void parse(@NotNull String in) {
        if (!in.startsWith(">>>")) {
            return;
        }
        @NotNull String str = in.substring(3);
        System.out.println("Matching " + str);
        @NotNull Matcher m = regex.matcher(str);
        if (!m.matches()) {
            System.out.println("Doesn't match");
            return;
        }
        String fn = m.group(1);
        System.out.println(fn);
        @NotNull Object args = m.group(2).split(",");
        System.out.println(System.currentTimeMillis());
        @NotNull Invocable inv = (Invocable) engine;
        try {
            inv.invokeFunction(fn, args);
        } catch (@NotNull ScriptException | NoSuchMethodException ex) {
            Logger.getLogger(ExternalConsole.class.getName()).log(Level.SEVERE, null, ex);
        }
        System.out.println(System.currentTimeMillis());
    }

    void setOut(@Nullable OutputStream s) {
        input.setEnabled(s != null);
        pw = new PrintWriter(s, true);
        engine.getContext().setWriter(pw);
    }

    /**
     * @return the engine
     */
    ScriptEngine getEngine() {
        return engine;
    }
}

File: src/main/kotlin/com/timepath/hl2/ExternalConsole.kt
package com.timepath.hl2

import com.timepath.Logger
import com.timepath.with
import java.awt.BorderLayout
import java.awt.Dimension
import java.awt.Font
import java.awt.event.ActionListener
import java.awt.event.WindowAdapter
import java.awt.event.WindowEvent
import java.io.*
import java.net.InetAddress
import java.net.Socket
import java.util.Date
import java.util.logging.Level
import javax.script.Invocable
import javax.script.ScriptEngine
import javax.script.ScriptEngineManager
import javax.script.ScriptException
import javax.swing.*
import javax.swing.text.DefaultCaret
import kotlin.concurrent.thread
import kotlin.platform.platformStatic

/**
 * http://www.perkin.org.uk/posts/how-to-fix-stdio-buffering.html
 */
public open class ExternalConsole protected constructor() : JFrame() {
    private val input: JTextField
    protected val output: JTextArea
    var engine = initScriptEngine()
        private set
    private var pw: PrintWriter? = null
    private var sock: Socket? = null

    init {
        output = JTextArea() with {
            setFont(Font("Monospaced", Font.PLAIN, 15))
            setEnabled(false)
            getCaret() as DefaultCaret with {
                setUpdatePolicy(DefaultCaret.ALWAYS_UPDATE)
            }
        }
        getContentPane().add(JScrollPane(output) with {
            setHorizontalScrollBarPolicy(ScrollPaneConstants.HORIZONTAL_SCROLLBAR_AS_NEEDED)
            setVerticalScrollBarPolicy(ScrollPaneConstants.VERTICAL_SCROLLBAR_ALWAYS)
        }, BorderLayout.CENTER)
        input = JTextField() with {
            setEnabled(false)
            addActionListener(ActionListener {
                pw?.let { pw ->
                    output.append("] ")
                    pw.println(input.getText())
                    input.setText("")
                }
            })
        }
        getContentPane().add(input, BorderLayout.SOUTH) // TODO: work out better way of sending input
        setJMenuBar(JMenuBar() with {
            add(JMenu("File") with {
                add(JMenuItem("Reload script") with {
                    addActionListener { engine = initScriptEngine() }
                })
            })
        })
        setTitle("External console")
        // setAlwaysOnTop(true);
        // setUndecorated(true);
        setPreferredSize(Dimension(800, 600))
        addWindowListener(object : WindowAdapter() {
            override fun windowClosing(e: WindowEvent) {
                try {
                    sock?.close()
                } catch (ex: IOException) {
                    LOG.log(Level.SEVERE, { null }, ex)
                }
                dispose()
            }
        })
        pack()
    }

    private fun initScriptEngine(file: String = "extern.js"): ScriptEngine {
        return ScriptEngineManager().getEngineByExtension(file.substringAfterLast("."))!! with {
            val bindings = createBindings()
            bindings["loadTime"] = Date()
            getContext().setWriter(pw)
            try {
                eval(FileReader(file))
            } catch (ex: ScriptException) {
                LOG.log(Level.SEVERE, { null }, ex)
            } catch (ex: FileNotFoundException) {
                LOG.log(Level.SEVERE, { null }, ex)
            }
        }
    }

    protected fun connect(port: Int) {
        sock = Socket(InetAddress.getByName(null), port)
        setIn(sock!!.getInputStream())
        setOut(sock!!.getOutputStream())
    }

    fun setIn(s: InputStream?) {
        output.setEnabled(s != null)
        thread {
            try {
                s?.reader()?.forEachLine {
                    update(it)
                }
            } catch (ex: IOException) {
                LOG.log(Level.SEVERE, { null }, ex)
            }
        }
    }

    fun update(str: String) {
        parse(str)
        appendOutput(str)
    }

    private fun appendOutput(str: String) = output.append("$str\n")

    protected open fun parse(lines: String) {
        if (!lines.startsWith(">>>")) {
            return
        }
        val str = lines.substring(3)
        System.out.println("Matching $str")
        val m = regex.match(str)
        if (m == null) {
            System.out.println("Doesn't match")
            return
        }
        val fn = m.groups[1]!!.value
        System.out.println(fn)
        val args = m.groups[2]!!.value.splitBy(",")
        System.out.println(System.currentTimeMillis())
        try {
            (engine as Invocable).invokeFunction(fn, args)
        } catch (ex: ScriptException) {
            LOG.log(Level.SEVERE, { null }, ex)
        } catch (ex: NoSuchMethodException) {
            LOG.log(Level.SEVERE, { null }, ex)
        }
        System.out.println(System.currentTimeMillis())
    }

    fun setOut(s: OutputStream?) {
        input.setEnabled(s != null)
        pw = PrintWriter(s!!, true)
        engine.getContext().setWriter(pw)
    }

    companion object {

        private val LOG = Logger()
        private val regex = "(\\S+)\\s*[(]\\s*(\\S*)\\s*[)].*".toRegex()

        public fun exec(cmd: String, breakline: CharSequence?): String = StringBuilder {
            try {
                val sock = Socket(InetAddress.getByName(null), 12345)
                val pw = PrintWriter(sock.getOutputStream(), true)
                val br = sock.getInputStream().bufferedReader()
                pw.println(cmd)
                if (breakline != null) {
                    br.readLine() // first line is echoed
                    for (line in br.lines()) {
                        appendln(line)
                        if (breakline in line) break
                    }
                }
                sock.close()
            } catch (ex: IOException) {
                LOG.log(Level.SEVERE, { null }, ex)
            }
        }.toString()

        public platformStatic fun main(args: Array<String>) {
            ExternalConsole() let {
                it.connect(12345)
                it.setVisible(true)
            }
        }
    }
}
